plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoIcebergCatalog.java

```diff
@@ -0,0 +1,286 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.iceberg;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.collect.ImmutableMap;
+import io.trino.plugin.base.CatalogName;
+import io.trino.plugin.hive.HdfsEnvironment;
+import io.trino.spi.TrinoException;
+import io.trino.spi.connector.ConnectorMaterializedViewDefinition;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.connector.SchemaNotFoundException;
+import io.trino.spi.security.TrinoPrincipal;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.Path;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.Transaction;
+import org.apache.iceberg.UpdateProperties;
+import org.apache.iceberg.catalog.Catalog;
+import org.apache.iceberg.catalog.Namespace;
+import org.apache.iceberg.catalog.SupportsNamespaces;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.exceptions.NamespaceNotEmptyException;
+import org.apache.iceberg.exceptions.NoSuchNamespaceException;
+import org.apache.iceberg.util.PropertyUtil;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+
+import static io.trino.plugin.hive.HiveMetadata.TABLE_COMMENT;
+import static io.trino.plugin.iceberg.IcebergConfig.convertToCatalogProperties;
+import static io.trino.plugin.iceberg.IcebergUtil.schemaFromTableId;
+import static io.trino.spi.StandardErrorCode.NOT_SUPPORTED;
+import static io.trino.spi.StandardErrorCode.SCHEMA_NOT_EMPTY;
+import static java.util.Objects.requireNonNull;
+import static org.apache.iceberg.CatalogUtil.loadCatalog;
+
+/**
+ * Implementation of {@link TrinoCatalog} that allows plugin an Iceberg {@link Catalog}.
+ * Because Iceberg catalog is single tenant, each session should use a new catalog instance to enforce access control.
+ * However, it might be able to reuse the same catalog instance based on the authZ info of a session.
+ * See {@link #getSessionCacheKey(ConnectorSession)} and {@link #getSessionProperties(ConnectorSession)} for more details.
+ */
+public class TrinoIcebergCatalog
+        implements TrinoCatalog
+{
+    private static final Map<String, String> EMPTY_SESSION_MAP = ImmutableMap.of();
+
+    private final String catalogImpl;
+    private final String catalogName;
+    private final Map<String, String> catalogProperties;
+    private final String warehouse;
+    private final HdfsEnvironment hdfsEnvironment;
+    private final Cache<String, Catalog> catalogCache;
```

Comment:
The intention was to first have this basic implementation, and we can iterate on efficiency later. Is that a possible way to go just to avoid too much discussion leading into that area?